runtime/auth/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/awssigning/AbstractAwsChunked.kt

```diff
@@ -0,0 +1,227 @@
+) : SdkByteReadChannel by chan {
+    companion object {
+        const val CHUNK_SIZE_BYTES: Int = 65536
+    }
+
+    override val isClosedForRead: Boolean
+        get() = chan.isClosedForRead && (chunk == null || chunkOffset >= chunk!!.size)
+
+    var chunk: ByteArray? = null
+    var chunkOffset: Int = 0
+    var hasLastChunkBeenSent: Boolean = false
+
+    /**
+     * Returns all the bytes remaining in the underlying data source, up to [limit].
+     * @return a [ByteArray] containing at most [limit] bytes. it may contain fewer if there are less than [limit] bytes
+     * remaining in the data source.
+     */
+    override suspend fun readRemaining(limit: Int): ByteArray {
+        if (chunk == null || chunkOffset >= chunk!!.size) { chunk = getNextChunk() }
+
+        var bytesWritten = 0
+        var bytes = byteArrayOf()
+        while (bytesWritten != limit) {
+            val numBytesToWrite: Int = minOf(limit - bytesWritten, chunk!!.size - chunkOffset)
+
+            bytes += chunk!!.slice(chunkOffset until chunkOffset + numBytesToWrite)
+
+            bytesWritten += numBytesToWrite
+            chunkOffset += numBytesToWrite
+
+            // read a new chunk. this handles the case where we consumed the whole chunk but still have not sent `limit` bytes
+            if (chunkOffset >= chunk!!.size) {
+                chunk = getNextChunk()
+                if (chunk == null) { break } // we've exhausted all remaining bytes
+            }
+        }
+
+        return bytes
+    }
+
+    /**
+     * Writes [length] bytes to [sink], starting [offset] bytes from the beginning. If [length] bytes are not available in
+     * the source data, the call will fail with an [IllegalArgumentException].
+     *
+     * @param sink the destination [ByteArray] to write to
+     * @param offset the number of bytes in [sink] to skip before beginning to write
+     * @param length the number of bytes to write to [sink]
+     * @throws IllegalArgumentException when illegal [offset] and [length] arguments are passed
+     * @throws RuntimeException when the source data is exhausted before [length] bytes are written to [sink]
+     */
+    override suspend fun readFully(sink: ByteArray, offset: Int, length: Int) {
+        require(offset >= 0) { "Invalid read: offset must be positive:  $offset" }
+        require(offset + length <= sink.size) { "Invalid read: offset + length should be less than the destination size: $offset + $length < ${sink.size}" }
+
+        if (chunk == null || chunkOffset >= chunk!!.size) { chunk = getNextChunk() }
+
+        var bytesWritten = 0
+        while (bytesWritten != length) {
+            if (chunk == null) { throw RuntimeException("Invalid read: unable to fully read $length bytes. missing ${length - bytesWritten} bytes.") }
+
+            val numBytesToWrite: Int = minOf(length, chunk!!.size - chunkOffset)
+
+            val bytes = chunk!!.slice(chunkOffset until chunkOffset + numBytesToWrite).toByteArray()
```

Comment:
Comment: Note you could be using `ByteArray.sliceArray` which will give you back a `ByteArray` directly (no conversion to/from `List<Byte>`. A better solution though is to just directly call `chunk.copyInto(sink, ...)` without the extra allocation. [copyInto](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/copy-into.html) accepts the destination offset as well as the source start/end parameters